kf/nfconfig.go

```
@@ -942,3 +943,367 @@ func getHostInterfaces() (map[string]bool, error) {
+	default:
+		return fmt.Errorf("unknown direction type")
+	}
+
+	for e := bpfList.Front(); e != nil; e = e.Next() {
+		data := e.Value.(*BPF)
+		if data.Program.Name == bpfProg.Name {
+			log.Warn().Msgf("%v is already running on %v iface and in %v direction ", data.Program.Name, ifaceName, direction)
+			return nil
+		}
+
+		if data.Program.SeqID == bpfProg.SeqID {
+			log.Warn().Msgf("duplicate seq Id detected for %v in direction %v", data.Program.Name, direction)
+			return nil
+		}
+	}
+	for e := bpfList.Front(); e != nil; e = e.Next() {
+		data := e.Value.(*BPF)
+		if data.Program.SeqID > bpfProg.SeqID {
+			bpf := NewBpfProgram(c.ctx, *bpfProg, c.HostConfig)
+			tmpBPF := bpfList.InsertBefore(bpf, e)
+			if err := c.DownloadAndStartBPFProgram(tmpBPF, ifaceName, direction); err != nil {
+				return fmt.Errorf("failed to download and start eBPF program %s version %s iface %s direction %s", bpfProg.Name, bpfProg.Version, ifaceName, direction)
+			}
+
+			if tmpBPF.Next() != nil {
+				if err := c.LinkBPFPrograms(tmpBPF.Value.(*BPF), tmpBPF.Next().Value.(*BPF)); err != nil {
+					log.Error().Err(err).Msg("AddAndStartBPF - failed LinkBPFPrograms after InsertBefore element to with next prog")
+					return fmt.Errorf("AddAndStartBPFProg - failed LinkBPFPrograms after InsertBefore element to with next prog %v", err)
+				}
+			}
+			return nil
+		}
+	}
+
+	// insert at the end
+	if err := c.PushBackAndStartBPF(bpfProg, ifaceName, direction); err != nil {
+		return fmt.Errorf("failed to push back and start eBPF Program %s version %s iface %s direction %s", bpfProg.Name, bpfProg.Version, ifaceName, direction)
+	}
+
+	return nil
+}
+
+// AddProgramWithoutChaining : add eBPF program on given interface when chaining is not enabled
+func (c *NFConfigs) AddProgramWithoutChaining(ifaceName string, bpfProgs *models.BPFPrograms) error {
+	if c.HostConfig.BpfChainingEnabled {
+		return nil
+	}
+	for _, bpfProg := range bpfProgs.XDPIngress {
```

I think we should drop the for loop logic here as if we are adding a program when chaining is disabled there can only be 1 added.  If bpfProgs > 1 we should just return an error as this is invalid operation that is being requested.

The same applies to the two other for loops below